docs/specs/ML.Net Sweepable API.md

```diff
@@ -0,0 +1,162 @@
+# AutoML.Net Sweepable API proposal
+## Overview
+Sweepable api allows mlnet users to create their own search space and pipeline for hyper-parameter optimization (HPO). It comes with three major part: [search space](#search-space), [sweepable estimator](#sweepable-estimator) and [tuner](#tuner). And all API lives under `Sweepable()` extension (for now).
+
+## search space
+Search space defines a range of hyper-parameter for tuner to search from. Sweepable API provides two way to create a search space.
+
+via attribute
+```csharp
+public class Option
+{
+    [Range(2, 32768, init: 2, logBase: true)]

```

Comment:
Do we need to have a reflection based solution to start? Or would the "weakly-typed" API solve most of the scenarios:

```C#
var ss = new SearchSpace();
ss.Add("WindowSize", new UniformIntOption(2, 32768, true, 2));
ss.Add("SeriesLength", new ChoiceOption(2,3,4));
ss.Add("UseSoftmax", new ChoiceOption(true, false));
ss.Add("AnotherOption", ss.Clone())
```

Reply:
The reason I ask is: I would not add multiple ways to do something right away. Instead, make the "core" thing first, and the simpler ones can be built later, if necessary.

Reply:
The reflection API (let's call it strong-typed correspond with "weakly-typed") is just a handy way for creating search space and it's built on top of "weakly-typed". So if we must make a choice at the beginning I would pick "weakly-typed" to implement/migrate first.

Reply:
### Resolved